### ISSUE TITLE
Cannot reference outer scopes from inside a `{% call %}` block.

### ISSUE THREAD

**Issue by carljm** [mozilla/nunjucks#679]
Originally reported by @jgerigmeyer at https://github.com/mozilla/nunjucks/pull/481#issuecomment-124220254 -- all of the below copied from his report:

Consider the following test case:

```
{% macro inner(class=none) %}
  <div class="{{ class }}">{{ caller() }}</div>
{% endmacro %}

{% macro outer(class=none) %}
  <div class="{{ class }}">
    {% call inner() %}
      {{ class }}
    {% endcall %}
  </div>
{% endmacro %}

{{ outer(class='outer') }}
```

Jinja2 outputs as expected:

```
<div class="outer">
  <div class="None">
    outer
  </div>
</div>
```

But nunjucks (with this PR) outputs:

```
<div class="outer">
  <div class="">
  </div>
</div>
```

**Comment by carljm** [comment]
This is related to #664; it has to do with how the frame stack for a `{% call %}` block is constructed.

**Comment by atheros** [comment]
This is a very annoying bug. You cannot take full advantage of call/caller feature without it.

Are there any plans for fixing this bug?

**Comment by ArmorDarks** [comment]
yeap, macros scopes are kinda buggy right now :(

@atheros as far as I remember, making call with `{% call %}` makes it work:

```jinja2
{% call outer(class='outer) %}{% endcall %}
```

Ugly, but better than nothing.

**Comment by atheros** [comment]
It doesn't work, the result is the same.

**Comment by ArmorDarks** [comment]
You're probably on Nunjucks 3.0.0? As far as I remember, in 3.0.0 it is impossible to pass external variable into macros at all, due to bug described here https://github.com/mozilla/nunjucks/issues/912

**Comment by atheros** [comment]
Yes, 3.0.0. And I cannot pass variables to call block, which pretty much makes call useless.

Anyway downgrading to 2.5.2 fixes the issue. Thanks for the hint @ArmorDarks 

Interestingly, set block in macros doesn't work there (bug #914), so it's an older issue.

**Comment by jgerigmeyer** [comment]
I note that this bug still exists in v3.2.0.
